### PR TITLE
Feature/s3 writes

### PR DESCRIPTION
**Pull request by samansmink**
# Overview
This PR adds S3 multipart uploads to DuckDB. Both the CSV and Parquet copy functions are supported. The uploads are implemented as filesystem writes in the existing S3FS class. Data written to s3fs is buffered in multiple buffers, one for each part of the multipart upload. Once a buffer is full, a thread is launched that uploads the buffer to S3 asynchronously. One of the goals of the multipart upload is that using multiple concurrent connections should be faster than a single. Another advantage of the multipart upload is that we can do streaming uploads. Finally, a retry mechanism is implemented for part upload request that fail. This means that the uploader is resilient to brief network interruptions/changes (timeout is configurable).

For now the writes have to be sequential. File appends are currently not working since this is not supported in S3. Also, once FileSync or Close are called on the FileHandle, the S3 file is finalised in S3 the file can no longer be written

During testing I was running into very sporadic but persistent httplib::Error::Read and httplib::Error::Connection errors on large reads. To resolve this, two measures are now in place: I have shared the connections between requests using keep-alive. Secondly, the read and write timeout are now increase to 30s for all requests. This seems to have resolved the issue as I haven't been able to reproduce the errors. Note that in this process the HTTPlib was updated to the latest version with the changes made in duckdb to it re-applied.

# Configuration
For the S3 uploader, we have to work with some limitations of AWS S3. The most important limitations/properties are:

- Each connection has a limited bandwidth
- Each file can be uploaded in maximally 10000 parts
- Each part must be of equal size, except the last one, which can be smaller.
- The minimum part size is 5MiB
- The maximum file size is 5TB
- Uploads are priced per request: uploading 1 file in 10000 parts is 10000x more expensive that uploading it at once.

Since we cannot cover all use cases with one configuration given the s3 limitations, there are 3 user defined settings that can be modified with the SET operation:

- **s3_uploader_max_filesize** determines the maximum filesize we can upload.
- **s3_uploader_max_parts_per_file** sets the maximum number of parts that will be used for a single file. 
- **s3_uploader_thread_limit** determines the amount of upload threads that can be launched globally. Each thread has its own upload buffer whose size is equal to max_filesize/max_parts

Through some basic experiments I've found the following settings to be a reasonable balance between performance, memory usage, and pricing.
- s3_uploader_max_filesize: 800GB 
- s3_uploader_max_parts_per_file: 10000 (aws default)
- s3_uploader_thread_limit: 50

These settings mean that:
- the part_size is 80MB
- ~12 500 requests are made per TB uploaded (costs: ~$0.0625/TB)
- max upload buffer memory is 50*80MB = 4GB (can be less if duckdb memory limit is set)

## Testing
For testing I added a simple s3 test server to the CI, s3rver (https://github.com/jamhall/s3rver). Also two new features are added to SQLLogicTestRunner: require-env and set. Require-env conditionally enables a test based on an environment variable. Set allows defining two parameters that were previously hard-coded: *ignore_error_messages* and *always_fail_error_messages*. Both parameters are comma separated list of values that, when found in an error message during the test, are respectively ignored and counted as a pass or not counted as a pass when encountered in a "statement error". 

## Benchmarks
For slow network connections like the one I have at home (~30Mbps) A single upload thread will max out the upload bandwith therefore basically all settings with reasonably sized part sizes will work fine.

When benchmarking on a machine on a faster network (~870 Mbps up measured), I expected the lowest part size to be the fastest by far, however, the upload speed to s3 seems to be less limited per connection than I expected. Below are the results for a simple copy query of the TPC-H Sf1 lineitem table to S3 in parquet format: ```COPY lineitem TO 's3://test-bucket-ceiveran/multipart/export_large.parquet' (FORMAT 'parquet');``` The size of the parquet file is 270MB. Results are averaged over 3 runs with a pretty big variance, so the differences between all but the last run are probably not significant.

Part size (MB)| AVG runtime (s)|Total speed of copy (Mbps)
-|-|-
5|7.927487333|282.7629873
10|9.329702333|240.2649002
20|10.210258|219.5439136
40|8.676809667|258.3438022
80|6.155621333|364.1549534
160|8.680137|258.2447719
320|12.00409333|186.7363022

The results show that the upload speed is not influenced much by the concurrency of the upload unless we reach the full size of the file. This suggests very little parallel uploads are actually required to reach max upload speed of very roughly ~280Mpbs.

## Some potential improvements that can still be made
While the CSV and Parquet s3 copy work well, we may later want to look into some of these improvements:
- Allow FileSync without Finalization of S3 File
- Support random writes in some degree (may be tricky due to deadlock potential)
- Support appends by using UploadPartCopy to copy the file inside S3, then append to the copy (https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html)

### CODE REVIEW

File: extension/httpfs/httpfs.cpp [hunk 1]
```diff
@@ -1,61 +1,153 @@
 #include "httpfs.hpp"
+
+#include "duckdb/common/file_opener.hpp"
+#include "duckdb/common/thread.hpp"
 #include "duckdb/function/scalar/strftime.hpp"
+
 #define CPPHTTPLIB_OPENSSL_SUPPORT
 #include "httplib.hpp"
 
+#include <atomic>
```
Comment (Mytherin) [inline extension/httpfs/httpfs.cpp:10]: `duckdb/common/atomic.hpp`

File: extension/httpfs/httpfs-extension.cpp [hunk 2]
```diff
@@ -10,21 +11,37 @@ namespace duckdb {
 static void LoadInternal(DatabaseInstance &instance) {
 	S3FileSystem::Verify(); // run some tests to see if all the hashes work out
 	auto &fs = instance.GetFileSystem();
+
 	fs.RegisterSubSystem(make_unique<HTTPFileSystem>());
-	fs.RegisterSubSystem(make_unique<S3FileSystem>());
+	fs.RegisterSubSystem(make_unique<S3FileSystem>(*instance.GetStorageManager().buffer_manager));
```
Comment (Mytherin) [inline extension/httpfs/httpfs-extension.cpp:16]: `BufferManager::GetBufferManager(*instance)`

Then the `storage_manager.hpp` include can also be removed.

File: extension/httpfs/include/httpfs.hpp [hunk 3]
```diff
@@ -6,37 +6,59 @@
...
-
+class Client;
+} // namespace duckdb_httplib_openssl
 namespace duckdb {
 
 using HeaderMap = unordered_map<string, string>;
 
-struct ResponseWrapper { /* avoid including httplib in header */
+// avoid including httplib in header
+struct ClientDeleter {
```
Comment (Mytherin) [inline extension/httpfs/include/httpfs.hpp:16]: A better way of doing this is to define the destructor of the object holding the unique_ptr in the `.cpp` file. That way this custom deleter is not required and the code can be simplified.

File: extension/httpfs/include/s3fs.hpp [hunk 4]
```diff
@@ -2,27 +2,157 @@
 
 #include "httpfs.hpp"
 #include "duckdb/common/file_opener.hpp"
+#include "duckdb/common/mutex.hpp"
+#include "duckdb/storage/buffer_manager.hpp"
+
+#include <atomic>
```
Comment (Mytherin) [inline extension/httpfs/include/s3fs.hpp:8]: `duckdb/common/atomic.hpp`

File: extension/httpfs/httpfs.cpp [hunk 5]
```diff
@@ -1,61 +1,153 @@
...
+	proto_host_port = url.substr(0, slash_pos);
 
-	auto path = url.substr(slash_pos);
+	path = url.substr(slash_pos);
 	if (path.empty()) {
 		throw std::runtime_error("URL needs to contain a path");
 	}
+	auto &hfs = (HTTPFileHandle &)handle;
+	if (!hfs.http_client) {
+		handle.http_client = unique_ptr<duckdb_httplib_openssl::Client, ClientDeleter>(
```
Comment (Mytherin) [inline extension/httpfs/httpfs.cpp:32]: `make_unique` can be used after we remove the ClientDeleter

File: extension/httpfs/httpfs.cpp [hunk 6]
```diff
@@ -1,61 +1,153 @@
...
-	cli.enable_server_certificate_verification(false);
+	return {timeout};
+}
+
+unique_ptr<ResponseWrapper> HTTPFileSystem::PostRequest(FileHandle &handle, string url, HeaderMap header_map,
+                                                        unique_ptr<char[]> &buffer_out, idx_t &buffer_out_len,
+                                                        char *buffer_in, idx_t buffer_in_len) {
+	auto &hfs = (HTTPFileHandle &)handle;
+	auto path = init_request(hfs, url);
+	unique_ptr<duckdb_httplib_openssl::Headers> headers = initialize_http_headers(header_map);
```
Comment (Mytherin) [inline extension/httpfs/httpfs.cpp:70]: `auto`

File: extension/httpfs/httpfs.cpp [hunk 7]
```diff
@@ -1,61 +1,153 @@
...
+		throw std::runtime_error("HTTP PUT error on '" + url + "' (Error code " + std::to_string((int)res.error()) +
+		                         ")");
+	}
+	return make_unique<ResponseWrapper>(res.value());
+}
+
+unique_ptr<ResponseWrapper> HTTPFileSystem::HeadRequest(FileHandle &handle, string url, HeaderMap header_map) {
+	auto &hfs = (HTTPFileHandle &)handle;
+	auto path = init_request(hfs, url);
+	unique_ptr<duckdb_httplib_openssl::Headers> headers = initialize_http_headers(header_map);
```
Comment (Mytherin) [inline extension/httpfs/httpfs.cpp:118]: `auto`

File: extension/httpfs/httpfs.cpp [hunk 8]
```diff
@@ -1,61 +1,153 @@
...
+		throw std::runtime_error("HTTP HEAD error on '" + url + "' (Error code " + std::to_string((int)res.error()) +
+		                         ")");
+	}
+	return make_unique<ResponseWrapper>(res.value());
+}
+unique_ptr<ResponseWrapper> HTTPFileSystem::GetRangeRequest(FileHandle &handle, string url, HeaderMap header_map,
+                                                            idx_t file_offset, char *buffer_out, idx_t buffer_out_len) {
+	auto &hfs = (HTTPFileHandle &)handle;
+	auto path = init_request(hfs, url);
+	unique_ptr<duckdb_httplib_openssl::Headers> headers = initialize_http_headers(header_map);
```
Comment (Mytherin) [inline extension/httpfs/httpfs.cpp:131]: `auto`

File: extension/httpfs/s3fs.cpp [hunk 9]
```diff
@@ -106,48 +58,372 @@ static HeaderMap create_s3_get_header(std::string url, std::string query, std::s
...
+S3ConfigParams S3ConfigParams::ReadFrom(FileOpener *opener) {
+	uint64_t uploader_max_filesize;
+	uint64_t max_parts_per_file;
+	uint64_t max_upload_threads;
+	Value value;
+
+	if (opener->TryGetCurrentSetting("s3_uploader_max_filesize", value)) {
+		uploader_max_filesize = DBConfig::ParseMemoryLimit(value.GetValue<string>());
+	} else {
+		uploader_max_filesize = 800000000000; // 800GB
```
Comment (Mytherin) [inline extension/httpfs/s3fs.cpp:152]: Perhaps define these constants as a `constexpr` in the `S3ConfigParams` object?

File: extension/httpfs/httpfs.cpp [hunk 10]
```diff
@@ -1,61 +1,153 @@
...
-	    "bytes=" + std::to_string(file_offset) + "-" + std::to_string(file_offset + buffer_len - 1);
-	// printf("%s(%llu, %llu)\n", method.c_str(), file_offset, buffer_len);
+	return make_unique<ResponseWrapper>(res.value());
+}
+
+unique_ptr<ResponseWrapper> HTTPFileSystem::PutRequest(FileHandle &handle, string url, HeaderMap header_map,
+                                                       char *buffer_in, idx_t buffer_in_len) {
+	auto &hfs = (HTTPFileHandle &)handle;
+	auto path = init_request(hfs, url);
+	unique_ptr<duckdb_httplib_openssl::Headers> headers = initialize_http_headers(header_map);
```
Comment (Mytherin) [inline extension/httpfs/httpfs.cpp:105]: `auto`

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great. Some comments:

**Comment by Mytherin** [comment]
There are also some failures in the `Windows Extensions` test:

```
2022-02-11T11:16:42.7116438Z s3fs.obj : error LNK2019: unresolved external symbol "public: __cdecl duckdb::BufferHandle::~BufferHandle(void)" (??1BufferHandle@duckdb@@QEAA@XZ) referenced in function "public: __cdecl std::unique_ptr<class duckdb::BufferHandle,struct std::default_delete<class duckdb::BufferHandle> >::~unique_ptr<class duckdb::BufferHandle,struct std::default_delete<class duckdb::BufferHandle> >(void)" (??1?$unique_ptr@VBufferHandle@duckdb@@U?$default_delete@VBufferHandle@duckdb@@@std@@@std@@QEAA@XZ) [D:\a\duckdb\duckdb\build\release\extension\httpfs\httpfs_loadable_extension.vcxproj]
2022-02-11T11:16:42.7168202Z s3fs.obj : error LNK2019: unresolved external symbol "public: unsigned char * __cdecl duckdb::BufferHandle::Ptr(void)" (?Ptr@BufferHandle@duckdb@@QEAAPEAEXZ) referenced in function "public: static void __cdecl duckdb::S3FileSystem::UploadBuffer(class duckdb::S3FileHandle &,class std::shared_ptr<class duckdb::S3WriteBuffer>)" (?UploadBuffer@S3FileSystem@duckdb@@SAXAEAVS3FileHandle@2@V?$shared_ptr@VS3WriteBuffer@duckdb@@@std@@@Z) [D:\a\duckdb\duckdb\build\release\extension\httpfs\httpfs_loadable_extension.vcxproj]
2022-02-11T11:16:42.7170853Z s3fs.obj : error LNK2019: unresolved external symbol "public: class std::unique_ptr<class duckdb::BufferHandle,struct std::default_delete<class duckdb::BufferHandle> > __cdecl duckdb::BufferManager::Allocate(unsigned __int64)" (?Allocate@BufferManager@duckdb@@QEAA?AV?$unique_ptr@VBufferHandle@duckdb@@U?$default_delete@VBufferHandle@duckdb@@@std@@@std@@_K@Z) referenced in function "protected: class std::shared_ptr<class duckdb::S3WriteBuffer> __cdecl duckdb::S3FileSystem::GetBuffer(class duckdb::S3FileHandle &,unsigned short)" (?GetBuffer@S3FileSystem@duckdb@@IEAA?AV?$shared_ptr@VS3WriteBuffer@duckdb@@@std@@AEAVS3FileHandle@2@G@Z) [D:\a\duckdb\duckdb\build\release\extension\httpfs\httpfs_loadable_extension.vcxproj]
```

This is caused by the methods named there not being tagged with `DUCKDB_API`. This breaks now because additional methods from the main library are used within the HTTPFS extension, which need the `DUCKDB_API` tag to be able to resolve the linking on Windows.

Could you have a look at fixing those as well?

**Comment by samansmink** [comment]
yea I was about to ask about the unresolved symbols, I'll add the tags!

**Comment by samansmink** [comment]
After some testing I found out yesterday why the upload speed was a bit lower than expected, the upload paralellism wasnt working properly due to the connection reuse feature that I added. Now I've disabled connection reuse for the put and post request and the upload speed now reaches close to full upload bandwidth on the Scilens rocks machine I tested on with ~860Mbps up.  As a baseline I also uploaded the files with the AWS cli. 

Method | Source | FileType | Size (MB) | Runtime (s) | Upload speed (Mbps)
-- | -- | -- | -- | -- | --
AWS CLI | TPCH SF10 Lineitem | Parquet (snappy) | 2942 | 28.86 | 815.45
AWS CLI | TPCH SF4 Lineitem | CSV | 3066 | 32.47 | 755.50
DuckDB S3 (Default settings) | TPCH SF10 Lineitem | Parquet (snappy) | 2942 | 42.70 | 551.18
DuckDB S3 (Default settings) | TPCH SF4 Lineitem | CSV | 3066 | 31.41 | 780.89
DuckDB S3 (Minimum partsize) | TPCH SF10 Lineitem | Parquet (snappy) | 2942 | 43.93 | 535.75
DuckDB S3 (Minimum partsize) | TPCH SF4 Lineitem | CSV | 3066 | 32.04 | 765.52
DuckDB S3 (Default settings) | TPCH SF4 Lineitem | Parquet (uncompressed) | 2899 | 28.61 | 810.66

We can see that upload speeds match the AWS CLI for CSV uploads and uncompressed parquet. For the default compressed parquet the compression appears to form a bit of a bottleneck. Note that compression will still yield a much higher effective upload rate:

Method | Source | FileType | Size (MB) | Runtime (s) | Upload speed (Mbps) | Effective upload (Mbps)
-- | -- | -- | -- | -- | -- | --
DuckDB S3 (Default settings) | TPCH SF4 Lineitem | Parquet (uncompressed) | 2899 | 28.61 | 810.66 | 810.66
DuckDB S3 (Default settings) | TPCH SF4 Lineitem | Parquet (snappy) | 1177 | 18.99 | 495.91 | 1,221.45

**Comment by Mytherin** [comment]
Indeed I think these failed tests are unrelated to this PR. The ODBC test should be fixed now. The NodeJS test seems to have broken recently also in the master - not sure yet what the cause there is.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**